src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1406,25 +1407,48 @@ data Insertion a
 -- are no guarantees that they will behave the same.
 --
 -- @since 2.2.7
-data LockingKind
-    = ForUpdate
+data LockingKind where
+    ForUpdate :: LockingKind
       -- ^ @FOR UPDATE@ syntax.  Supported by MySQL, Oracle and
       -- PostgreSQL.
       --
       -- @since 2.2.7
-    | ForUpdateSkipLocked
+    ForUpdateSkipLocked :: LockingKind
       -- ^ @FOR UPDATE SKIP LOCKED@ syntax.  Supported by MySQL, Oracle and
       -- PostgreSQL.
       --
       -- @since 2.2.7
-    | ForShare
+    ForShare :: LockingKind
       -- ^ @FOR SHARE@ syntax.  Supported by PostgreSQL.
       --
       -- @since 2.2.7
-    | LockInShareMode
+    LockInShareMode :: LockingKind
       -- ^ @LOCK IN SHARE MODE@ syntax.  Supported by MySQL.
       --
       -- @since 2.2.7
+    ForUpdateOfSkipLocked :: [LockableEntity] -> LockingKind
+      -- ^ @FOR UPDATE OF tablename SKIP LOCKED@ syntax. Supported by MySQL, and PostgreSQL
+      --
+      -- @since 3.6.0.0
+
+-- | Wraps table type entities for use with LockingKind. 
+--
+-- Example use:
+--
+-- @
+-- select $ do
+--     (p :& bp) <- from $ 
+--         table @Person
+--         `innerJoin` table @BlogPost
+--             `on` do
+--                 \(p :& bp) -> p ^. PersonId ==. b ^. BlogPostAuthorId
+--     forUpdateOfSkipLocked [LockableEntity p,LockableEntity b]
+--     return p
+-- @
+--
+-- @since 3.6.0.0
+data LockableEntity where
+  LockableEntity :: PersistEntity val => (SqlExpr (Entity val)) -> LockableEntity
```

Comment:
I think this will not make it into the released one - the idiom of using a type class argument that can accept tuples and `:&` is very nice syntactically. It also works on `SqlExpr (Maybe (Entity val))` which you get back from a `leftJoin`.

```haskell
class LockableEntity a where
    toLockableEntity :: a -> Whatever

instance LockableEntity (SqlExpr (Entity rec)) 

instance LockableEntity (SqlExpr (Maybe (Entity rec)))

instance (LockableEntity a, LocakbleEntity b) => LockableEntity (a :& b)

-- etc
```

These instances allow you to write, eg, 

```haskell
-- one table
t <- from $ table @Foo
forUpdateOfSkipLocked t

t :& s <- from $ table @Foo `leftJoin` table @Bar `on` ...
forUpdateOfSkipLocked s

all@(t :& s) <- from $ table @Foo `leftJoin` table @Bar `on` ...
forUpdateOfSkipLocked all
```

Reply:
I think we actually don't want (SqlExpr (Maybe (Entity val)) to be available for the locking clause because of the error:

```
FOR UPDATE cannot be applied to the nullable side of an outer join
```

I'm not sure there's another case where we want to be able to do this with a nullable side since we would want the error to be caught by the type checker anyway. (ie we would only want to be able to do that for `inner joins` or the non nullable side of outer joins)

I don't feel super strongly about this though so happy to make the change if you prefer it (Just for clarification what is the `Whatever` type retyrned by `toLockableEntity`?)

Reply:
That's a great catch!

In that case, we can even write a `TypeError` instance for `SqlExpr (Maybe (Entity rec))` 😄 

I'm happy to implement this after we get this merged. I think I will cut a branch for 3.5 soon and use that for backporting fixes and get 3.6 to an unreleased master.